src/goto-synthesizer/cegis_verifier.cpp

```diff
@@ -537,49 +581,86 @@ optionalt<cext> cegis_verifiert::verify()
       violation_type = cext::violation_typet::cex_not_hold_upon_entry;
     }
 
+    // The violation is an assignable check.
+    if(property_it.second.description.find("assignable") != std::string::npos)
+    {
+      violation_type = cext::violation_typet::cex_assignable;
+    }
+
+    // Compute the cause loop---the loop for which we synthesize loop contracts,
+    // and the counterexample.
+
+    // If the violation is an assignable check, we synthesize assigns targets.
+    // In the case, cause loops are all loops the violation is in. We keep
+    // adding the new assigns target to the most-inner loop that does not
+    // contain the new target until the assignable violation is resolved.
+
+    // For other cases, we synthesize loop invariant clauses. We synthesize
+    // invariants for one loop at a time. So we return only the first cause loop
+    // although there can be multiple ones.
+
+    log.debug() << "Start to compute cause loop ids." << messaget::eom;
+
     // The loop which could be the cause of the violation.
-    // We say a loop is the cause loop if the violated predicate is dependent
-    // upon the write set of the loop.
-    optionalt<loop_idt> cause_loop_id = get_cause_loop_id(
-      checker->get_traces()[property_it.first], property_it.second.pc);
+    std::list<loop_idt> cause_loop_ids;
+
+    // Doing assigns-synthesis or invariant-synthesis
+    if(violation_type == cext::violation_typet::cex_assignable)
+    {
+      checked_pointer = static_cast<const exprt &>(
+        property_it.second.pc->condition().find(ID_checked_assigns));
+      cause_loop_ids = get_cause_loop_id_for_assigns(trace);
+      cext result(violation_type);
+      result.cause_loop_ids = cause_loop_ids;
+      result.checked_pointer = checked_pointer;
```

Comment:
I'd put `cext result(violation_type);` and would then avoid the use of temporaries for you can safely do the assignments to `result.cause_loop_ids` and `result.checked_pointer` directly. I'd then suggest to inspect whether there is any use of `checked_pointer` and `cause_loop_ids` left, or whether these can be removed entirely. If they have to stay, then place the declarations as close as possible to the actual use.